### PR TITLE
FS-1018 assessment store initial

### PR DESCRIPTION
**Pull request by harryyo**
Added assessment store api with sqlite/postgresql db backend

### CODE REVIEW

File: config/__init__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from os import environ
+
+FLASK_ENV = environ.get("FLASK_ENV")
+
+if FLASK_ENV == "development":
+    from config.environments.development import DevelopmentConfig as Config
+elif FLASK_ENV == "dev":
+    from config.environments.dev import DevConfig as Config
```
Comment (Adam-W1) [inline config/__init__.py:8]: are these meant to be two different envs?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,12 +1,13 @@
 repos:
 -   repo: https://github.com/ambv/black
-    rev: 22.3.0
+    rev: 21.12b0
```
Comment (gidsg) [inline .pre-commit-config.yaml:3]: why are we downgrading this Can we revert?

File: config/utils.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                if service.get("name") == name:
+                    return service
+            raise Exception(f"Service name '{name}' not found")
+        raise Exception(f"Service group '{group_key}' not found")
+
+    def get_service_credentials_value(
+        self, group_key: str, name: str, key: str
+    ):
+        service = self.get_service_by_name(group_key, name)
+        return service.get("credentials").get(key)
```
Comment (gidsg) [inline config/utils.py:29]: Can we just remove this until it is needed?

File: db/migrations/versions/0a08ce61dc30_assessments_table.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+"""assessments table
+
+Revision ID: 0a08ce61dc30
```
Comment (gidsg) [inline db/migrations/versions/0a08ce61dc30_assessments_table.py:3]: Can we use a more human readable/orderable revision ID e.g. `0001`?

File: app.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    )
+
+    # This is needed to access the running app's environment config
+    # outside the request context using env.config.get("VARIABLE_NAME")
+    env.init_app(flask_app)
+
+    from db import db, migrate
+
+    db.init_app(flask_app)
+    migrate.init_app(flask_app, db, directory="db/migrations")
```
Comment (gidsg) [inline app.py:36]: I think we may need to separate this out as a separate script to run on PaaS? Does this mean the migrations will run every time the app starts?

**Comment by gidsg** [comment]
Have added fix so that PaaS can connect to the database, and added blocking script to the deploy command so that database migrations run on PaaS before app starts.

File: license.md [hunk 6]
```diff
@@ -1,6 +1,4 @@
-MIT License
-
-Copyright (c) 2022 Crown Copyright (Department for Levelling Up, Housing and Communities)
+Copyright (c) 2022
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
```
Comment (gidsg) [inline license.md:3]: Can we revert this change please?

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -7,18 +7,19 @@ repos:
       args:
         - --line-length=79
         - --experimental-string-processing
```
Comment (SentientHousePlant) [inline .pre-commit-config.yaml:9]: 🔧 This isn't needed in the newest black version...it is no longer experimental.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,112 @@
...
+    python3 -m pip install --upgrade pip && pip install -r requirements-dev.txt
+
+NOTE: The psycopg2 package (required for PostgreSQL) can sometimes have difficulty installing on certain environments
+related to libssl.
+If you experience difficulties on macOS this might be required:
+
+    export LDFLAGS="-L/usr/local/opt/openssl/lib"
+
+or see other resolutions on [StackOverflow](https://stackoverflow.com/questions/11365619/psycopg2-installation-error-library-not-loaded-libssl-dylib)
+### Initialise and upgrade database
```
Comment (gio-karunakaran) [inline README.md:46]: 💯 very nice read me! helpful db instructions here!

File: api/assessments/routes.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+class AssessmentsView(AssessmentMethods, MethodView):
+    def search(self):
+        """
+        GET /assessments endpoint
+        :return: Json Response
+        """
+        return Response(
+            json.dumps(self.assessments(as_json=True)),
+            mimetype="application/json",
+        )
```
Comment (SentientHousePlant) [inline api/assessments/routes.py:21]: 💯 Nice to have this wrapped up in the response object...we should all try to use this...

File: api/assessments/routes.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        If the assessment_id matches a valid assessment in the db
+        return the assessment
+        If no matching valid link is found returns a 404 error message
+        :param assessment_id: id of the assessment
+        :return: 200 assessment JSON / 404 Error
+        """
+        try:
+            assessment = self.get_by_id(assessment_id)
+        except AssessmentError as e:
+            return error_response(404, e.message)
```
Comment (SentientHousePlant) [inline api/assessments/routes.py:35]: ❤️ ❤️ Love the custom errors, will dovetail nicely with the new logging we will have soon!

File: api/assessments/routes.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        :return: a json of the assessment created (or an error of failure)
+        """
+        application_id = request.get_json().get("applicationId")
+        if not application_id:
+            return error_response(400, "Application ID is required")
+
+        try:
+            new_assessment = self.register_application(application_id)
+        except AssessmentError as e:
+            return error_response(401, e.message)
```
Comment (SentientHousePlant) [inline api/assessments/routes.py:51]: 🤔 ⛏️ Could we have a newline between the try-except loop and the return statement?

File: app.py (hunk 5 above)
Comment (SentientHousePlant) [reply app.py:36]: It looks like this only initialises command line options to the flask app - see [here](https://flask-migrate.readthedocs.io/en/latest/). Might be worth adding this as a comment or to the readme.

File: db/migrations/README [hunk 12]
```diff
@@ -0,0 +1 @@
+Single-database configuration for Flask.
```
Comment (SentientHousePlant) [inline db/migrations/README:1]: 🔧 Copy over information about DBs from root README at least :)

File: db/migrations/script.py.mako [hunk 13]
```diff
@@ -0,0 +1,24 @@
+"""${message}
```
Comment (SentientHousePlant) [inline db/migrations/script.py.mako:1]: 🤔 Could you explain the usage of this file please? :)

File: db/models/assessment.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    )
+    application_id = db.Column(db.Text(), index=True, unique=True)
+
+    def __repr__(self):
+        return "<Assessment {} for Application {}>".format(
+            self.id, self.application_id
+        )
+
+    @property
+    def as_json(self):
```
Comment (SentientHousePlant) [inline db/models/assessment.py:22]: ⛏️ How about just "json" since it's a property

File: db/models/assessment.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    id = db.Column(
+        "id",
+        db.Text(),
+        default=lambda: str(uuid.uuid4()),
+        primary_key=True,
+    )
+    application_id = db.Column(db.Text(), index=True, unique=True)
+
+    def __repr__(self):
+        return "<Assessment {} for Application {}>".format(
```
Comment (SentientHousePlant) [inline db/models/assessment.py:17]: 🤔 Typically reprs mirror how an object instantiated, but that could be complicated here...something to think about. Lovely use of a dunder method though 💯 (f string instead of format....? :P)

File: openapi/api.yml [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+paths:
+  /assessments:
+    get:
+      tags:
+        - assessments
+      summary: Search assessments
+      description: List all assessments
+      operationId: api.AssessmentsView.search
+      responses:
+        200:
```
Comment (SentientHousePlant) [inline openapi/api.yml:19]: 🔧 Let's add an example.

File: openapi/api.yml [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+          content:
+            application/json:
+              schema:
+                $ref: 'components.yml#/components/schemas/Assessment'
+        401:
+          description: ERROR - Could not create assessment
+          content:
+            application/json:
+              schema:
+                $ref: 'components.yml#/components/schemas/GeneralError'
```
Comment (SentientHousePlant) [inline openapi/api.yml:54]: 🔧 Examples?

File: openapi/components.yml [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+      properties:
+        applicationId:
+          type: string
+    Assessment:
+      type: object
+      properties:
+        id:
+          type: string
+        applicationId:
+          type: string
```
Comment (SentientHousePlant) [inline openapi/components.yml:24]: 🔧 Examples?

File: pyproject.toml [hunk 19]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "funding-service-design-assessment-store"
+version = "0.1.1"
+description = "The funding service design assessment store for the DLUHC."
+authors = ["Version One"]
+license = "MIT License"
+
+[tool.black]
+line-length = 79
+experimental-string-processing = 1
```
Comment (SentientHousePlant) [inline pyproject.toml:10]: 🔧 not needed anymore

File: pyproject.toml [hunk 20]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "funding-service-design-assessment-store"
+version = "0.1.1"
+description = "The funding service design assessment store for the DLUHC."
+authors = ["Version One"]
```
Comment (SentientHousePlant) [inline pyproject.toml:5]: 🔧 Add dhluc

File: tests/helpers.py [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+
+    Args:
+        test_client: A flask test client
+        endpoint (str): The GET request endpoint
+        expected_data: The content we expect to find
+        exclude_paths: DeepDiff dict paths to exclude from diff errors
+        exclude_regex_paths: DeepDiff dict regex paths
+            to exclude from diff errors
+
+    """
```
Comment (SentientHousePlant) [inline tests/helpers.py:25]: 💯 These docstrings are giving me life.

File: config/environments/development.py [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+    FUND_STORE_API_HOST = "fund_store"
+    ROUND_STORE_API_HOST = "round_store"
+    NOTIFICATION_SERVICE_HOST = "notification_service"
+
+    # Security
+    FORCE_HTTPS = False
+
+    """
+    Database
+    """
```
Comment (SentientHousePlant) [inline config/environments/development.py:26]: 😠 Change to comments.

File: tests/mocks/flask_test_client.py [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+    Creates the test client we will be using to test the responses
+    from our app, this is a test fixture.
+    :return: A flask test client.
+    """
+
+    with create_app(testing=True).app_context() as app_context:
+        SqliteTestDB.create()
+        with app_context.app.test_client() as test_client:
+            yield test_client
+        SqliteTestDB.remove()
```
Comment (SentientHousePlant) [inline tests/mocks/flask_test_client.py:18]: 💯 Beautiful fixture.

**Review by SentientHousePlant — CHANGES_REQUESTED** [review]
🍰 Great stuff Harry, just some small pick-y stuff to fix! Lovely test work.

File: config/__init__.py (hunk 1 above)
Comment (harryyo) [reply config/__init__.py:8]: Yep exactly

File: .github/workflows/codeql-analysis.yml [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+name: "CodeQL"
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    # The branches below must be a subset of the branches above
+    branches: [ main ]
+  schedule:
+    - cron: '44 9 * * 2'
```
[the remark this answers is not in the document]
Comment (harryyo) [reply .github/workflows/codeql-analysis.yml:21]: No - this was set up previously I think by Cary

File: README.md (hunk 8 above)
Comment (harryyo) [reply README.md:46]: Glad helpful!

File: api/assessments/routes.py (hunk 9 above)
Comment (harryyo) [reply api/assessments/routes.py:21]: 👍

File: api/assessments/routes.py (hunk 10 above)
Comment (harryyo) [reply api/assessments/routes.py:35]: 👌

File: app.py (hunk 5 above)
Comment (harryyo) [reply app.py:36]: Comments add to clarify

File: db/migrations/README (hunk 12 above)
Comment (harryyo) [reply db/migrations/README:1]: This README is automatically generated by Flask-Migrate so don't have any control over it!

File: db/migrations/script.py.mako (hunk 13 above)
Comment (harryyo) [reply db/migrations/script.py.mako:1]: Again auto-generated by Flask-Migrate so part of migrations management

File: db/models/assessment.py (hunk 15 above)
Comment (harryyo) [reply db/models/assessment.py:17]: Obv f string - thanks, done

File: pyproject.toml (hunk 20 above)
Comment (harryyo) [reply pyproject.toml:5]: Added

File: tests/mocks/flask_test_client.py (hunk 23 above)
Comment (harryyo) [reply tests/mocks/flask_test_client.py:18]: Too kind

File: openapi/components.yml (hunk 18 above)
Comment (harryyo) [reply openapi/components.yml:24]: Added

File: openapi/api.yml (hunk 17 above)
Comment (harryyo) [reply openapi/api.yml:54]: Added

File: license.md (hunk 6 above)
Comment (harryyo) [reply license.md:3]: Reverted

File: db/models/assessment.py (hunk 14 above)
Comment (harryyo) [reply db/models/assessment.py:22]: converted to normal method as maybe that's the more standard convention for a serialisation operation

File: config/utils.py (hunk 3 above)
Comment (harryyo) [reply config/utils.py:29]: Removed

File: api/assessments/routes.py (hunk 11 above)
Comment (harryyo) [reply api/assessments/routes.py:51]: Added, much nicer

File: db/models/assessment.py (hunk 14 above)
Comment (SentientHousePlant) [reply db/models/assessment.py:19]: 💯
